### PR TITLE
Add Cloud SQL MySQL support

### LINKED ISSUE #7146: MySQL for GCP Cloud SQL support
### What

Teleport database access currently doesn't support GCP-hosted MySQL databases.

The reason is that we use IAM authentication with databases which MySQL engine doesn't support (only Postgres does). The only auth method Cloud SQL MySQL supports is username/password. There's also no indication from GCP folks when or whether IAM support for MySQL will become available.

As such, we might need to come up with alternative way to authenticate to GCP-hosted MySQL databases.

### How

GCP has an API for updating database user's password:

https://cloud.google.com/sql/docs/mysql/create-manage-users#changing_a_user_password
https://pkg.go.dev/google.golang.org/api@v0.47.0/sqladmin/v1beta4#UsersService.Update
https://cloud.google.com/sql/docs/mysql/admin-api/rest/v1beta4/users/update

See if we can use it to implement short-lived throw-away tokens for database users - i.e. update the user's password to a random value before each connect and use it as a password to connect.

One thing to keep in mind is how to handle clients connecting concurrently as the same database account? 

### Why

To support MySQL in GCP.

### Workaround

None.

### PR DESCRIPTION
**Pull request by r0mant**
Adds support for connecting to MySQL Cloud SQL databases with database access.

MySQL in Cloud SQL does not support IAM authentication, so instead we're using GCP Cloud SQL Admin API to emulate one-time passwords: before each connect we reset the requested database user's password to a generated token and then connect with it. See linked ticket for more details on the used APIs, if interested.

This approach means that we have to serialize all connections to the same database/user to avoid races with clients updating the password - esp. keeping in mind HA scenario where there may be multiple database services proxying the same database instance.

The nice property about this approach though is that it can be used with regular database users as opposed to Postgres IAM auth where you have to connect with service accounts.

I've also refactored the way database service handles creating cloud clients a bit along the way, to make it a bit cleaner, ensure better separation of concerns and also to make it easier in future to introduce dynamically added/removed databases - previously it would only initialize all clients upon startup, now it initializes required clients on demand and caches them.

Closes https://github.com/gravitational/teleport/issues/7146.

### CODE REVIEW

**Review by Joerger — APPROVED** [review]

**Review by xacrimon — APPROVED** [review]
LGTM but please beware that I lack experience with most of the stuff here and the Go MySQL stuff so some things may go over my head without me noticing.

**Review by smallinsky — APPROVED** [review]

File: lib/srv/db/common/auth.go [hunk 1]
```diff
@@ -172,6 +183,53 @@ func (a *dbAuth) GetCloudSQLAuthToken(ctx context.Context, sessionCtx *Session)
...
+	retryCtx, cancel := context.WithTimeout(ctx, defaults.DatabaseConnectTimeout)
+	defer cancel()
+	err = retry.For(retryCtx, func() error {
+		return a.updateCloudSQLUser(ctx, sessionCtx, gcpCloudSQL, &sqladmin.User{
+			Password: token,
+		})
+	}, func(err error) bool {
+		e, ok := trace.Unwrap(err).(*googleapi.Error)
+		return ok && e.Code == http.StatusConflict // We only want to retry on 409.
+	})
```
[the remark this answers is not in the document]
Comment (r0mant) [reply lib/srv/db/common/auth.go]: Yeah good call. I was thinking about this approach too and it's cleaner indeed. Updated.

File: lib/srv/db/common/cloud.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+	}
+	c.mtx.RUnlock()
+	return c.initGCPSQLAdminClient(ctx)
+}
+
+// Closes closes all initialized clients.
+func (c *cloudClients) Close() error {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	var errors []error
```
[the remark this answers is not in the document]
Comment (r0mant) [reply lib/srv/db/common/cloud.go]: The idea was that we may need to close other clients in future here as well, but I guess you're right that currently there's just a single error.

**Review by russjones — APPROVED** [review]
Bot.

**Pull request merged**